### ISSUE TITLE
How to run or install this driver

### ISSUE THREAD

**Issue by frustrul** [mmm444/eie-pro-linux#1]
Hi, 

I'm not a totaly linux freak, far from, but i know the basics, maybe just basics basic, however, even with audio on linux i am new. (Though I am definitly a big linux liker)

So I come from windows, and like to work with reaper, and  know I am doubting to give it another shot on linux. 

Abvious, I also have the AKAI eie pro. 

Can I try this driver, how do I install it and simpliest check it works? 

Clearly there is no spellcheck on here, so sorry for my english.

I was thinking to try Ardour3 as a daw, but as it seem I can not even run it, It blocks at creating new session in the begining, 
anyhows, 
It might be also cool for others if you also get some text on it on how to try/install this driver. 

Kind greetings,
Sam.

**Comment by fraunos** [comment]
Yes, i'm the same as you. I've been looking for a way to ask this question here, but i'm also github noob (how come i didn't come up with writing an issue..), so... You're not alone! Waiting eagerly for reply.

**Comment by fraunos** [comment]
Hm. I tried using
$ make

but it returns some errors, specifically

make -C /lib/modules/3.13.0-37-generic/build M=/home/michal/Pobrane/eie-pro-linux-master modules
make[1]: Entering directory `/usr/src/linux-headers-3.13.0-37-generic'
  CC [M]  /home/michal/Pobrane/eie-pro-linux-master/eie-pro.o
/home/michal/Pobrane/eie-pro-linux-master/eie-pro.c: In function ‘eie_probe’:
/home/michal/Pobrane/eie-pro-linux-master/eie-pro.c:1233:2: error: implicit declaration of function ‘snd_card_new’ [-Werror=implicit-function-declaration]
  err = snd_card_new(&interface_to_usbdev(interface)->dev, index[card_index], id[card_index], THIS_MODULE,
  ^
cc1: some warnings being treated as errors
make[2]: *** [/home/michal/Pobrane/eie-pro-linux-master/eie-pro.o] Error 1
make[1]: *** [_module_/home/michal/Pobrane/eie-pro-linux-master] Error 2
make[1]: Leaving directory`/usr/src/linux-headers-3.13.0-37-generic'
make: **\* [default] Error 2

I'm not sure if it's problem with my system or with the source code, but eh we have to wait anyway.

**Comment by mmm444** [comment]
Thanks for your feedback. Sorry that it took me so long to reply. I missed the notification.

I do not recommend using the driver even for testing at the moment. It simply does not work correctly and has to be fixed.

I will try to look at it in the upcoming days. Stay tuned. I will close this issue and add some basic build/usage instructions to the readme when it is ready at least for testing. Thanks for your patience.

@fraunos From the error message I see that you have 3.13 kernel but at least 3.15 is required - the `snd_card_new` function was introduced there. I was developing against 3.17.

**Comment by bartjan2005** [comment]
Would love to help testing but i'm both a linux and code rookie..

At least another enthusiast!

**Comment by LoranKey** [comment]
Yet another enthusiast !

I'm very interested by this project !
I'd like to help testing.

I tried : 

> #uname -r
> 3.18.20-1-MANJARO
> 
> #make
> cc -I/usr/include/libusb-1.0  -Wall -g -O2   -c -o pokus.o pokus.c
> cc   pokus.o  -lusb-1.0 -lsndfile  -lm -o pokus
> ./r
> Makefile:5 : la recette pour la cible « run » a échouée
> make: **\* [run] Erreur 124

**Comment by 59Panhead** [comment]
I am returning to Linux (Fedora 23) after being away from it for many years.  So, I'm very new to this environment.

I have the Akai EIE 16 bit version, and would really like to use it on Linux.  I haven't even tried it on Windows.  So, I'm watching this thread closely.  Thanks for the fine work.

**Comment by fraunos** [comment]
From what i know Akai EIE (not Pro) uses USB 1, so it doesn't need any special drivers (it's class compliant? i'm not sure if this is the right phrase), it's plug and play. But try it first. You will just need to install qjackctl or similar program and connect some DAW to it. There are many tutorials on the web, so it won't be a problem :)

**Comment by Aisyk** [comment]
Hello,

I planned to buy this Akai EIE and i need some returns of users using it. I'm on Elementary OS Freya (4.2 kernel).
My needs are for the 4 audio in (need to record in the same time).

Thanks for your response !

**Comment by mmm444** [comment]
Hello,

the driver works to some extent but it is not suitable for normal use. There are several severe bugs in it. For example the driver behaves differently on second opening of the device and there are some troubles with small buffer sizes and so on. The project is currently "on hold" and I would like to return to hack on it but I know it will not happen until summer. 

Thanks for everyone's interest.

**Comment by MatTaylorSharrow** [comment]
Hi mmm444,

I'd like to try to help to progress development of this driver.   I don't have lots of device driver experience but I'm fairly comfortable with C.  Do you have a todo list of jobs that should be worked on next?  What is required to take this driver closer to a stable driver?  

I've compiled the driver and I'm using it on my workstation and laptop.  The basic functionality works well, but there are lots of clicks and pops during many scenarios.  

I'm going to fork and see what I can do.  Any help/advice/documentation etc would be very gratefully received.

Thanks

**Comment by mmm444** [comment]
Hi @MatTaylorSharrow,

I am sorry for the late reply. You caught me on holiday. First of all big thanks for your effort.

> Do you have a todo list of jobs that should be worked on next?

Sorry, no. Only the TODOs in the code. But from a quick look over them they do no seem to be very relevant. :(

> What is required to take this driver closer to a stable driver?

Rethink and rework the way the driver communicates with the ALSA kernel streams API.

The USB side of the driver is IMHO functional. The URBs are allocated, sent and received properly. MIDI seems to be working. What is wrong is the interface to the ALSA linux kernel audio streams API. On the EIE Pro you always start the capture and playback streams simultaneously with the same bitrate. In the ALSA world the streams are separate. The driver has to start both streams internally whenever ALSA opens one and join the other once ALSA opens it as well. When an error happens in one direction you additionally have to deal with it in the other. The unmanaged complexity of the states that arise from this situation is (perhaps with some stream initialization details) the main source of problems. As far as I am concerned the best way to solve it is to rethink the details of the life cycle of the streams and then write them in code.

Sorry for the vague description. It has been a long time since I got to work on this driver and I don't remember the details.

I plan to spend some time working on this again in the last week of August. I hope I can push it a little bit further.

Did you make any progress meanwhile?

Thanks

**Comment by Thibo06** [comment]
Hi mmm444,
I'm on Ubuntu-Studio and will try to connect my EIE-Pro using your driver. I will tell you my troubles as I understand that the version is not stable (dual-stream management).
Are you still working on this issue ?
Thanks

**Comment by natohutch** [comment]
Hello,
I also have an EIE Pro and if you need any help testing would be keen to do so.
Are there any updates on your progress?
Thanks

**Comment by aum7** [comment]
another eie pro (love it, lol) owner here, would love to help with development, but other than testing (fedora returnee here, lol), i poses no sufficient coding knowledge.
detailed instructions how to setup this driver would also be beneficial, and dependencies etc. too.

best regards,

simon
s love nia

**Comment by cgw21** [comment]
It would be interesting reverse engineering the current working Mac drivers with ghidra, to see if that would help with the development. Might provide a good framework for continuing this project. If I get some time, I'll try to lay the initial ground work for this going forward.

**Comment by aum7** [comment]
cgw21, just don't get caught in the act (lol), we need you here.
if i got this right, we need alsa on linux for eie pro to work? since it is usb-connected.
go, cgw21, go go go! lol.

have fun,

simon

**Comment by cgw21** [comment]
Yes also will most likely be needed or at least be useful in getting it work how we all expect.

**Comment by sunjar2020** [comment]
Great project. I also have a EIE Pro, I will keep the focus on this project, and hope to use it on my Linux one day. I currently use Ubuntu 16.04 (my laptop) and 20.04(my PC).

Thanks for your work.

**Comment by naringas** [comment]
Hi, I just got this working with relative ease.

I confirm this works fine for playback. I may test audio input later. 

Here's what I did to get it working (manjaro linux (would also work for arch linux)):

1. Install basic develompent tools (`base-devel` group)
2. Install linux headers for my current running kernel (`packman -S linux54-headers`)
3. run `make` in the cloned git repository
4. compress the succesfully compiled module `xz *.ko`
5. copy (with sudo) the file into `/lib/modules/extramodules-...`
6. run (sudo) `depmod`
7. and finally `sudo modprobe eie_pro`

I am not sure if modprobe is even necessary. Nor have I figured out if I need anything else to make it autoload on reboot.

**Comment by jota-pfmarques** [comment]
Hello, 

I've run into some bad luck with my computer, and a friend gave me a lenovo that he had completely installed with Linux (Ubuntu), wiping Windows completely out).

The thing is I make music and my interface (EIE PRO) doesn't seem to work. I tried to install yours, but I'm not managing to.

I'm getting a little desperate, because i can't afford another interface and I need it for work.

Is there any chance you can guide me through the installation?
That would really be a lifesaver.

Thank you so much,
João Pedro

**Comment by robbyleh** [comment]
@mmm444 @all contrubutors: I've got the driver running on Linux Mint  , yeah. 
NAME="Linux Mint"
VERSION="22.3 (Zena)"
ID=linuxmint
ID_LIKE="ubuntu debian"
PRETTY_NAME="Linux Mint 22.3"
VERSION_ID="22.3"

reaper freeze in first attempts, finally I had to deactivate IEI for the Input, and it is working fine.
 

<img width="558" height="314" alt="Image" src="https://github.com/user-attachments/assets/f0a4fccd-9440-4177-bf3a-8582f65dfac8" />

Thank you very much for your effort here, so I can go ahead using some old hardware for my drum monitoring at home.

apt install build-essential linux-headers-$(uname -r)
make
cp eie_pro.ko /lib/modules/$(uname -r)/kernel/drivers/usb/
secure boot in UEFI bios abschalten
reboot
depmod
modprobe eie_pro
aplay -l
working ...
